cmd/query/app/querysvc/adjuster/hash.go

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package adjuster
+
+import (
+	"fmt"
+	"hash/fnv"
+
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	"github.com/jaegertracing/jaeger/internal/jptrace"
+)
+
+var _ Adjuster = (*SpanHashDeduper)(nil)
+
+// SpanHash creates an adjuster that deduplicates spans by removing all but one span
+// with the same hash code. This is particularly useful for scenarios where spans
+// may be duplicated during archival, such as with ElasticSearch archival.
+//
+// The hash code is generated by serializing the span into protobuf bytes and applying
+// the FNV hashing algorithm to the serialized data.
+//
+// To ensure consistent hash codes, this adjuster should be executed after
+// SortAttributesAndEvents, which normalizes the order of collections within the span.
+func SpanHash() SpanHashDeduper {
+	return SpanHashDeduper{
+		marshaler: &ptrace.ProtoMarshaler{},
+	}
+}
+
+type SpanHashDeduper struct {
+	marshaler ptrace.Marshaler
+}
+
+func (s *SpanHashDeduper) Adjust(traces ptrace.Traces) {
+	spansByHash := make(map[uint64]ptrace.Span)
+	resourceSpans := traces.ResourceSpans()
+	for i := 0; i < resourceSpans.Len(); i++ {
+		hashTrace := ptrace.NewTraces()
+		rs := resourceSpans.At(i)
+		hashResourceSpan := hashTrace.ResourceSpans().AppendEmpty()
+		rs.Resource().Attributes().CopyTo(hashResourceSpan.Resource().Attributes())
+		scopeSpans := rs.ScopeSpans()
+		hashScopeSpan := hashResourceSpan.ScopeSpans().AppendEmpty()
```

Comment:
hard to grok due to ordering and naming

```suggestion
		rs := resourceSpans.At(i)
		scopeSpans := rs.ScopeSpans()

		hashTrace := ptrace.NewTraces()
		hashResourceSpans := hashTrace.ResourceSpans().AppendEmpty()
		hashScopeSpans := hashResourceSpan.ScopeSpans().AppendEmpty()
		hashSpan := hashScopeSpans.Spans().AppendEmpty()
		rs.Resource().Attributes().CopyTo(hashResourceSpan.Resource().Attributes())
```